### ISSUE TITLE
@tool script object added to scene do not always get accurately recorded into .tscn

### ISSUE THREAD

**Issue by this-is-envy** [godotengine/godot#95646]
### Tested versions

- found in `v4.2.2.stable.official` [15073afe3]

### System information

Godot v4.2.2.stable - Windows 10.0.19045 - Vulkan (Forward+)

### Issue description

# Context
I'm building a plugin to add configured versions of a complex scene into our level scene.

# Issue
Under some circumstances the added objects are not faithfully saved into the `.tscn` even though the Scene Tree appears correct in editor.

~~The editor tab does not show `(*)` indicated unsaved changes and even if I make changes then force save the it doesn't save the addon-added objects.~~ Per conversation below this is because I'm not using the EditorUndoRedoPlugin. I added that into the workflow and now get the `(*)` badge until save but it doesn't impact the issue of incorrectly persisted scenes.

If you close the scene and reopen or play the scene some configuration is lost.

# Expectation
What I see in the editor will be saved into the tscn and persist into play or through an unload/load cycle.

### Steps to reproduce

I'm condensing my situation so this is at best an approximation of things but that _should_ be sufficient to break things

### Setup
1. Have scene that will be manipulated: `GreyboxObject.tscn`

```gdscript
@tool

class_name GreyboxObject
extends Node2D

@export_category("Functionality")
@export var can_block_movement: bool = true: # it's important that the default here is set to true
	get:
		return can_block_movement
	set(value):
		can_block_movement = value
		# on-update code elided

func _ready() -> void:
	if Engine.is_editor_hint():
		get_parent().set_editable_instance(self, true)

```

2. At some point when setting you're GreyboxObject up in editor decide that you don't want `can_block_movement` to default to true and change `can_block_movement` to false. In `GreyboxObject.tscn` in the editor this now shows up as a non-default value, e.g., it can be reverted

3. Have an addon that creates `GreyboxObject`s  and configures it
```gdscript
	...
	var obj: GreyboxObject = preload(GREYBOX_OBJECT_SCENE).instantiate()
	obj.name = name
	_objects_parent.add_child(obj)
	# the owner of the new greybox object is the level (Object's parent's parent)
	# by convention this should always be the edited scene
	obj.owner = _objects_parent.get_parent()
	obj.can_block_movement = collides
```

### Repro
At this point if I use the addon code to create a new greybox object in `SomeLevel.tscn` that should collide it'll get added to the scene and correctly configured -- basically can_block_movement triggers adding `RigidBody2D` etc so that you can configure colliders -- but it's impossible to get the object to be saved into the `SomeLevel.tscn` with `can_block_movement=true`.

Without that can_block_movement annotation loading the GreyboxObject results in incorrect state relative to what you see in the editor immediately after adding it and having it report as "saved"

### tl;dr
I had three layers of values.

1. Code: true (not the primitive's default)
2. sceneA using the above code: false (resets to primitive's default)
3. sceneB adding sceneA: true

Basically because 1 and 3 match the resulting scene isn't detecting the value as dirty/needing to be saved.

### Minimal reproduction project (MRP)

TBD, I tried to include sufficient information above but will also try to carve out time to add a MRP as well

**Comment by AThousandShips** [comment]
For updates to the scene saved state to work you need to use undo/redo, just adding a node to the scene or manipulating it isn't enough, see [here](https://docs.godotengine.org/en/stable/classes/class_editorundoredomanager.html#class-editorundoredomanager) for that class

**Comment by AThousandShips** [comment]
Please try setting the owner to the edited scene, with [`EditorInterface.get_edited_scene_root()`](https://docs.godotengine.org/en/stable/classes/class_editorinterface.html#class-editorinterface-method-get-edited-scene-root)

**Comment by this-is-envy** [comment]
## Parentage
> Please try setting the owner to the edited scene, with [`EditorInterface.get_edited_scene_root()`](https://docs.godotengine.org/en/stable/classes/class_editorinterface.html#class-editorinterface-method-get-edited-scene-root)

That's essentially what this bit does

```gdscript
	# the parent of the new greybox object is the level (Object's parent's parent)
	obj.owner = _objects_parent.get_parent()
```

The scene tree looks like this:

```
EditedScene
└── _objects_parent
    └── object_being_added
```

## Undo/Redo
Just to be clear: Is the suggestion that changes to an object will not be persisted unless they're registered via UndoRedo? That seems wildly complicated tbh. If that's the case IMO it should 100% be surfaced in the "how to write an editor plugin" docs; it doesn't seem discoverable at the moment unless you're just reading the class descriptions for all EditorXYZ classes.

In any event I'm reading through the EditorUndoRedo stuff rn. Thanks for the pointer.

**Comment by AThousandShips** [comment]
Not persisted, but the `(*)` is based on the undo/redo system

Also please separate two different concepts, parentage isn't the same as ownership, confusing the two in your report makes it harder to evaluate

**Comment by this-is-envy** [comment]
Ah, the `(*)` link to Undo/Redo call out is useful! Feels like a far less onerous system.

That said, I did go through and swap over to your suggested changes and got the same behavior when the defaults were layered in the way I described.

- I double checked that parent's parent was the correct owner (basically by convention in how i'm setting my scene up, sorry for being loose with wording there) and swapped over to using the edited scene for good measure to check that it didn't matter,
- I also used the UndoRedo path for one of the values to make sure that wasn't the issue.

The relevant code after your suggestions now looks like:

```gdscript
	var obj: GreyboxObject = preload(GREYBOX_OBJECT_SCENE).instantiate()
	obj.name = name
	_objects_parent.add_child(obj)
	# the parent of the new greybox object is the level (Object's parent's parent)
	# obj.owner = _objects_parent.get_parent()
	obj.owner = EditorInterface.get_edited_scene_root()
	var urm: EditorUndoRedoManager = _plugin_ref.get_undo_redo()
	urm.create_action("Set up Greybox Object")
	urm.add_do_property(obj, "can_block_movement", collides)
	urm.add_undo_property(obj, "can_block_movement", obj.can_block_movement)
	# obj.can_block_movement = collides
	urm.commit_action()
```

And still see the reported issue.

Unless there is some specific thing that's worth trying I think subsequent cycles I have for this report probably need to go to a MRP. Right now I can work around it by not having the code and the scene diverge defaults.

**Comment by AThousandShips** [comment]
This probably requires an MRP to pin down what specifically is going on yeah

**Comment by AThousandShips** [comment]
For the undo/redo you need to access the undo state of the current scene, which I think you fetch with `manager.get_history_undo_redo(manager.get_object_history_id(EditorInterface.get_edited_scene_root()))` but I'm not experienced with using this myself

**Comment by this-is-envy** [comment]
> For the undo/redo you need to access the undo state of the current scene, which I think you fetch with `manager.get_history_undo_redo(manager.get_object_history_id(EditorInterface.get_edited_scene_root()))` but I'm not experienced with using this myself

yea, idk. I only did a cursor read of the docs but it sounded like the EditorUndoRedoManager handles finding the appropriate history to merge with based on the context of the object being changed?

> EditorUndoRedoManager is a manager for [UndoRedo](https://docs.godotengine.org/en/stable/classes/class_undoredo.html#class-undoredo) objects associated with edited scenes. Each scene has its own undo history and EditorUndoRedoManager ensures that each action performed in the editor gets associated with a proper scene

In any event I think it's probably non-essential to the core of the bug so if/when i get around to MRP i'll probably just leave redo state out

**Comment by Juzek89** [comment]
![Image](https://github.com/user-attachments/assets/9c5457f2-ce70-4dac-9676-dd4dac656fb3)

I am interested to see how this goes, because I have the very same issue. 
The error message is " ERROR: res://addons/clutter_placer/clutter_dock.gd:132 - Invalid call function 'set_texture' in base 'Node3D (Visual)': Attempt to call a method on a placeholder instance. Check if the script is in tool mode."

which, I get, my little set_texture method is on something that isn't a tool, but it called it a placeholder instance, and it never puts that instance into the scene tree on the left, and it doesn't save with the scene. 

Also interesting to me is that my comments don't all show up, it stops after "Placing clutter", but in the view port  it shows the objects in the world, properly placed and everything. 

This is my first tool I've written, so I'm a total know nothing in this area. Just know I have what seems like the exact same issue. maybe my new context will help you all out.

**Comment by this-is-envy** [comment]
> [@this-is-envy](https://github.com/this-is-envy) Have you made any progress on the MRP or is this no longer an issue?

Per the comment in my original report -- I worked around it by changing the `@export` default to match the primitive default:

```
@export var can_block_movement: bool = true: # it's important that the default here is set to true
```

I also haven't retested with a newer build since we haven't switched version for this project or created a stand-alone MRP yet (sorry 🙇)

**Comment by AThousandShips** [comment]
Closing due to lack of response, if you provide the requested information this can be reopened
